### ISSUE TITLE
HTML sanitization is quite a bad idea

### ISSUE THREAD

**Issue by madmurphy** [doxygen/doxygen#7288]
Long ago I had created three aliases in my `doxyfile.conf`, with the purpose of listing some functions in a particular page named `fnoverview`.

```ini
ALIASES   = "startfnlist=\page fnoverview ^^<dl>" \
            "endfnlist=\page fnoverview ^^</dl>" \
            "utility{1}=\page fnoverview ^^<dt> `\1()` </dt><dd> \copybrief \1 </dd>"
```

The argumentless alias `startfnlist` would output a simple `<dl>`, the alias `utility{name-of-the-function}` would output a simple `<dt></dt><dd></dd>` pair, in the form of `<dt>name-of-the-function()</dt><dd>brief-of-the-function</dd>`, and finally the argumentless alias `endfnlist` would output a simple `</dl>`.

These aliases used to work fantastically until version 1.8.14, so that when I organized my C code in the following way:

```c
/**  @startfnlist  **/

/**  @utility{do_something_1}  **/
/**

    @brief      Do something 1
    @param      some_int        Blabla
    @return     Something

**/
int do_something_1 (int some_int) {

    // Do something 1

}


/**  @utility{do_something_2}  **/
/**

    @brief      Do something 2
    @param      some_int        Blabla
    @return     Something

**/
int do_something_2 (int some_int) {

    // Do something 1

}

/**  @endfnlist  **/
```

the following HTML code would be appended to the `fnoverview` page:

```html
<dl>
    <dt><code><a class="el" href="..." title="Do something 1">do_something_1()</a></code></dt>
    <dd>Do something 1</dd>
    <dt><code><a class="el" href="..." title="Do something 2">do_something_2()</a></code></dt>
    <dd>Do something 2</dd>
</dl>
```

The code above would only be _further added_ to the `fnoverview` page, without preventing these functions to be listed in the “File Reference” page in the first place (which is exactly what I wanted).

But unfortunately something happened in Doxygen's new versions, probably some sort of HTML sanitization, wherewith Doxygen now sees the `<dt>...</dt><dd>...</dd>` pairs as missing the `<dl>` parent, and replaces them with `<p>...</p><p>...</p>`, printing a series of warnings:

```
warning: Unexpected tag <dd> found
```

The problem is that the `<dl>` parent is not missing at all, it's just that Doxygen forgets about it. In fact the HTML code produced would be perfectly valid.

The whole situation comes from Doxygen's attempt to sanitize a language as complex as HTML automatically generated from sparse sources.

Not even PHP – which is _born_ for outputting HTML – tries to sanitize the HTML code generated, and not just because it's a lost battle, but because the power of automatically generated HTML relies exactly on the possibility of composing a series of invalid HTML fragments into a perfectly valid HTML tree.

Given that Doxygen is used by developers, who normally know what they do, given that Doxygen is incapable of managing a situation like the one described here, and given that trying to sanitize HTML is _a really bad idea in general_ that only adds complexity to an already complex project, wouldn't be better to remove this sort of random HTML sanitization?

--madmurphy

**Comment by albert-github** [comment]
I tried to reproduce the problem with 1.8.14 and 1.8.16.

With 1.8.14 I got the warning:
```
warning: @copybrief or @copydoc target 'do_something_1' not found
warning: @copybrief or @copydoc target 'do_something_2' not found
```
with 1.8.16 I got:
```
warning: @copybrief or @copydoc target 'do_something_1' not found
warning: @copybrief or @copydoc target 'do_something_2' not found
warning: expected <dt> tag but found <br> instead!
warning: Unexpected tag <dd> found
warning: Unexpected tag <dd> found
```

To be able to reproduce the problems in a consistent way with you (and dive deeper into it):
- Can you please attach a, small, self contained example (source+configuration file in a tar or zip) that allows us to reproduce the problem?  Please don't add external links as they might not be persistent.

**Comment by madmurphy** [comment]
Sure, [here](https://github.com/doxygen/doxygen/files/3676250/doxygen_test.zip) is a complete minimal example.

It is certainly a sanitization problem, because if I surround the `utility` alias within `<dl>...</dl>` Doxygen does what it is instructed to do, although it is not something very helpful.
```
utility{1}=\page fnoverview ^^<dl><dt> `\1()` </dt><dd> \copybrief \1 </dd></dl>
```

**Comment by albert-github** [comment]
When looking at the statement:
```
/**  @startfnlist  **/
```
Note the 2 spaces before the `**/`. 
The markdown definition (https://daringfireball.net/projects/markdown/syntax#p) gives:

>    When you do want to insert a break tag using Markdown, you end a line with two or more spaces, then type return.

this was not correctly implemented but fixed in the 1.8.16 release (pull request #7163).
So when you replace the 2 spaces after `startfnlist` with just 1 space everything works again as you expect. Same of course is valid for the other comment blocks as well.
Another way would be to set `MARKDOWN_SUPPORT=NO` but at that moment you would use quite a lot of other useful features (a.o. the backtick support).

A special note is that one gets also a number of messages like:
```
warning: multiple use of page label 'fnoverview', (other occurrence:...
```
as the way you use the `\page` command is not the intended way (but works in this case, doxygen does append the page texts).

I think the problem is usage.

**Comment by madmurphy** [comment]
Thank you, Albert, removing the two spaces fixed it!

I wonder why the other warnings though. If I want to insert a text into a page from another page, isn't `@page [NAME-OF-THE-PAGE]` the way to do it?

And if I may, I would like to add a comment about Doxygen's current implementation of HTML line breaks. The markdown definition clearly says that you need two spaces + a line break to get a `<br />` tag. The following example would fit perfectly into the definition:

```c
// “When you do want to insert a break tag using Markdown ...”
/**
@startfnlist  //<-- “... you end a line with two or more spaces, ...” (there are two spaces here)
**/ //<-- “... then type return.” (there is a line break before `**/`)
```

But, to be precise, my example wouldn't fit in it:

```c
/**  @startfnlist  **/ //<-- there is no line break after the two spaces
```

It is not just for being pedantic, it's also because it is quite common for programmers to push the two comment initializers `/**` and `**/` far away from their text, for graphical purposes. And to this I would then add that it seems non-standard to implement a line break there.

--madmurphy
